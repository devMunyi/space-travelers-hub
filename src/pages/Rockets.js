import React, { useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import {
  fetchRockets,
  bookedRocket,
  cancelledRocket,
} from '../Redux/actions/rockets';

const Rockets = () => {
  // set dispatch
  const dispatch = useDispatch();

  // fetch rockets on page mount
  useEffect(() => {
    dispatch(fetchRockets());
  }, [dispatch]);

  // retrieve rockets from store for rendering
  const rockets = useSelector((state) => state.rockets);

  // book rocket
  const bookRocket = (id) => dispatch(bookedRocket(id));

  // cancel rocket book
  const cancelRocket = (id) => dispatch(cancelledRocket(id));

  return (
    <div className="container pt-2 pb-5">
      {rockets &&
        rockets.map((rocket) => (
          <div key={rocket.id} className="row mb-4">
            <div className="col-md-3">
              <img
                className="img-fluid"
                src={rocket.flickr_images[0]}
                alt={rocket.rocket_name}
              />
            </div>
            <div className="col-md-9">
              <h5>{rocket.rocket_name} </h5>
              <p>
                {' '}
                {rocket.reserved && (
                  <span className="badge bg-success">Reserved</span>
                )}{' '}
                {rocket.description}
              </p>{' '}
              {rocket.reserved && (
                <button
                  onClick={() => cancelRocket(rocket.id)}
                  type="button"
                  className="btn btn-sm btn-outline-secondary"
                >
                  Cancel Reservation
                </button>
              )}
              {!rocket.reserved && (
                <button
                  onClick={() => bookRocket(rocket.id)}
                  type="button"
                  className="btn btn-sm btn-primary"
                >
                  Reserve Rocket
                </button>
              )}
            </div>
          </div>
        ))}
    </div>
  );
};

export default Rockets;
