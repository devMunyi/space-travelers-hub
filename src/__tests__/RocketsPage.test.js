import { screen, render } from '@testing-library/react';
import { rest } from 'msw';
import { setupServer } from 'msw/node';
import Rockets from '../pages/Rockets';
// import store from '../Redux/store';

// import { Provider } from 'react-redux';

const server = setupServer(
  rest.get('/url', (req, res, ctx) => res(ctx.json([{ id: 1, rocket_name: 'Falcon 1' }]))),
);

beforeAll(() => server.listen());
afterEach(() => server.resetHandlers());
afterAll(() => server.close());

describe('Rockets Page', () => {
  it('should render rockets page correctly', () => {
    // ARRANGE
    render(<Rockets />);

    // ACT
    const text = screen.findByRole('heading');

    // ASSERT
    expect(text).toHaveTextContent('Falcon 1');
  });
});
