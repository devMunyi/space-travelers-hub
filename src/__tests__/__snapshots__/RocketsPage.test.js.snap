// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders Component correctly 1`] = `
<div
  className="container pt-2 pb-5"
/>
`;
